modules/calib3d/src/ptsetreg.cpp

```diff
@@ -501,6 +501,269 @@ class Affine3DEstimatorCallback : public PointSetRegistrator::Callback
+
+        for(int i = 0; i < count; i++ )
+        {
+            const Point2f& f = from[i];
+            const Point2f& t = to[i];
+
+            double a = F[0]*f.x + F[1]*f.y + F[2] - t.x;
+            double b = F[3]*f.x + F[4]*f.y + F[5] - t.y;
+
+            errptr[i] = (float)(a*a + b*b);
+        }
+    }
+
+    bool checkSubset( InputArray _ms1, InputArray _ms2, int count ) const
+    {
+        const float threshold = 0.996f;
+        Mat ms1 = _ms1.getMat(), ms2 = _ms2.getMat();
+
+        for( int inp = 1; inp <= 2; inp++ )
+        {
+            int j, k, i = count - 1;
+            const Mat* msi = inp == 1 ? &ms1 : &ms2;
+            const Point2f* ptr = msi->ptr<Point2f>();
+
+            CV_Assert( count <= msi->rows );
+
+            // check that the i-th selected point does not belong
+            // to a line connecting some previously selected points
+            for(j = 0; j < i; ++j)
+            {
+                Point2f d1 = ptr[j] - ptr[i];
+                float n1 = d1.x*d1.x + d1.y*d1.y;
+
+                for(k = 0; k < j; ++k)
+                {
+                    Point2f d2 = ptr[k] - ptr[i];
+                    float denom = (d2.x*d2.x + d2.y*d2.y)*n1;
+                    float num = d1.x*d2.x + d1.y*d2.y;
+
+                    if( num*num > threshold*threshold*denom )
+                        return false;
+                }
+            }
+        }
+        return true;
+    }
+};
+
+class AffinePartial2DEstimatorCallback : public Affine2DEstimatorCallback
+{
+public:
+    int runKernel( InputArray _m1, InputArray _m2, OutputArray _model ) const
+    {
+        Mat m1 = _m1.getMat(), m2 = _m2.getMat();
+        const Point2f* from = m1.ptr<Point2f>();
+        const Point2f* to   = m2.ptr<Point2f>();
+
+        // for partial affine trasform we need only 2 points
+        const int N = 4;
+        double buf[N*N + N + N + 2*3];
+        Mat A(N, N, CV_64F, buf);
+        Mat B(N, 1, CV_64F, buf + N*N);
+        Mat X(N, 1, CV_64F, buf + N*N + N);
+        Mat M(2, 3, CV_64F, buf + N*N + N + N);
+        double* Adata = A.ptr<double>();
+        double* Bdata = B.ptr<double>();
+        double* Xdata = X.ptr<double>();
+        double* Mdata = M.ptr<double>();
+        A = Scalar::all(0);
+
+        for( int i = 0; i < (N/2); i++ )
+        {
+            Bdata[i*2] = to[i].x;
+            Bdata[i*2+1] = to[i].y;
+
+            double *aptr = Adata + i*2*N;
+            aptr[0] = from[i].x;
+            aptr[1] = -from[i].y;
+            aptr[2] = 1.0;
+            aptr[4] = from[i].y;
+            aptr[5] = from[i].x;
+            aptr[7] = 1.0;
+        }
+
+        solve(A, B, X, DECOMP_SVD);
+
+        // set model, rotation part is antisymmetric
+        Mdata[0] = Mdata[4] = Xdata[0];
+        Mdata[1] = -Xdata[1];
+        Mdata[2] = Xdata[2];
+        Mdata[3] = Xdata[1];
+        Mdata[5] = Xdata[3];
+        M.copyTo(_model);
+        return 1;
+    }
+};
```

Comment:
Nice catch. Thank you.

I also though about too close points when I was implementing it (because estimateRigidTransform checks that), but neither estimateAffine3D nor findHomography does that, so I think it's ok to assume that this is in fact not causing too much problems.